src/terminal/input/terminalInput.cpp

```diff
@@ -515,169 +162,442 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
-        // The below condition primitively ensures that we allow all common Alt+Ctrl combinations
-        // while preserving most of the functionality of Alt+Ctrl as a substitute for AltGr.
-        if (ctrlAltChar == UNICODE_SPACE || (ctrlAltChar > 0x40 && ctrlAltChar <= 0x5A))
+        return keySequence;
+    }
+
+    // If it's not in the key map, we'll use the UnicodeChar, if provided.
+    if (unicodeChar != 0)
+    {
+        // In the case of an AltGr key, we may still need to apply a Ctrl
+        // modifier to the char, either because both Ctrl keys were pressed,
+        // or we got a LeftCtrl that was distinctly separate from the RightAlt.
+        const auto bothCtrlsArePressed = WI_AreAllFlagsSet(controlKeyState, CTRL_PRESSED);
+        const auto rightAltIsPressed = WI_IsFlagSet(controlKeyState, RIGHT_ALT_PRESSED);
+        if (altGrIsPressed && (bothCtrlsArePressed || (rightAltIsPressed && leftCtrlIsReallyPressed)))
         {
-            // Pressing the control key causes all bits but the 5 least
-            // significant ones to be zeroed out (when using ASCII).
-            return _makeEscapedOutput(ctrlAltChar & 0b11111);
+            unicodeChar = _makeCtrlChar(unicodeChar);
         }
-
-        // Currently, when we're called with Alt+Ctrl+@, ch will be 0, since Ctrl+@ equals a null byte.
-        // VkKeyScanW(0) in turn returns the vkey for the null character (ASCII @).
-        // -> Use the vkey to determine if Ctrl+@ is being pressed and produce ^[^@.
-        if (ch == UNICODE_NULL && vkey == LOBYTE(OneCoreSafeVkKeyScanW(0)))
+        auto charSequence = _makeCharOutput(unicodeChar);
+        if (!charSequence.empty())
         {
-            return _makeEscapedOutput(L'\0');
+            // We may also need to apply an Alt prefix to the char sequence, but
+            // if this is an AltGr key, we only do so if both Alts are pressed.
+            const auto bothAltsArePressed = WI_AreAllFlagsSet(controlKeyState, ALT_PRESSED);
+            if (altGrIsPressed ? bothAltsArePressed : altIsPressed)
+            {
+                charSequence = _makeEscapedOutput(charSequence);
+            }
         }
+        return charSequence;
+    }
+
+    // If we don't have a UnicodeChar, we'll try and determine what the key
+    // would have transmitted without any Ctrl or Alt modifiers applied. But
+    // this only makes sense if there were actually modifiers pressed.
+    if (!altIsPressed && !ctrlIsPressed)
+    {
+        return MakeOutput({});
     }
 
-    // If a modifier key was pressed, then we need to try and send the modified sequence.
-    if (WI_IsAnyFlagSet(keyEvent.dwControlKeyState, MOD_PRESSED))
+    // We need the current keyboard layout and state to lookup the character
+    // that would be transmitted in that state (via the ToUnicodeEx API).
+    const auto hkl = GetKeyboardLayout(GetWindowThreadProcessId(GetForegroundWindow(), nullptr));

```

Comment:
This will get the keyboard layout of the current active window, which may not be desireable since each window can have a different desired keyboard layout. (In particular when using an IME - when you switch windows, the active layout in the taskbar will also change accordingly.)

Is there any reason to get an explicit HKL? When you pass null to `ToUnicodeEx` it'll grab the current thread's keyboard layout. Is that breaking something?

Reply:
If I don't do this, it doesn't pick up the active keyboard for Windows Terminal. For example if you open a tab with a US keyboard, and then switch to a French keyboard, some keys end up being incorrectly mapped.

I assumed it had something to do with the fact that this code is being executed in conhost rather than Windows Terminal. Conhost by itself doesn't have this problem (if I remember correctly).

I'm open to other suggestions if there's a better way to deal with this, but just setting HKL to null doesn't work.